common-lib/src/main/scala/com/gu/mediaservice/lib/auth/Authentication.scala

```diff
@@ -1,122 +1,105 @@
+    case BuiltInAuthService => List(Link("login", config.services.loginUriTemplate))
+    case ExternalLoginLink(link) => List(Link("login", link))
+  }
 
-  keyStore.scheduleUpdates(actorSystem.scheduler)
+  def unauthorised(errorMessage: String, throwable: Option[Throwable] = None): Future[Result] = {
+    logger.info(s"Authentication failure $errorMessage", throwable.orNull)
+    Future.successful(respondError(Unauthorized, "authentication-failure", "Authentication failure", loginLinks))
+  }
 
-  private val userValidationEmailDomain = config.stringOpt("panda.userDomain").getOrElse("guardian.co.uk")
+  def forbidden(errorMessage: String): Future[Result] = {
+    logger.info(s"User not authorised: $errorMessage")
+    Future.successful(respondError(Forbidden, "principal-not-authorised", "Principal not authorised", loginLinks))
+  }
 
-  override lazy val panDomainSettings = buildPandaSettings()
+  def expired(user: UserPrincipal): Future[Result] = {
+    logger.info(s"User token expired for ${user.email}, return 419")
+    Future.successful(respondError(new Status(419), errorKey = "authentication-expired", errorMessage = "User authentication token has expired", loginLinks))
+  }
 
-  final override def authCallbackUrl: String = s"${config.services.authBaseUri}/oauthCallback"
+  def authenticationStatus(requestHeader: RequestHeader) = {
+    def flushToken(resultWhenAbsent: Result): Result = {
+      providers.userProvider.flushToken.fold(resultWhenAbsent)(_(requestHeader, resultWhenAbsent))
+    }
 
-  override def invokeBlock[A](request: Request[A], block: Authentication.Request[A] => Future[Result]): Future[Result] = {
-    // Try to auth by API key, and failing that, with Panda
-    request.headers.get(Authentication.apiKeyHeaderName) match {
-      case Some(key) =>
-        keyStore.lookupIdentity(key) match {
-          case Some(apiKey) =>
-            logger.info(s"Using api key with name ${apiKey.identity} and tier ${apiKey.tier}", apiKey)
-            if (ApiAccessor.hasAccess(apiKey, request, config.services))
-              block(new AuthenticatedRequest(ApiKeyAccessor(apiKey), request))
-            else
-              Future.successful(ApiAccessor.unauthorizedResult)
-          case None => Future.successful(invalidApiKeyResult)
+    // Authenticate request. Try with API authenticator first and then with user authenticator
+    providers.apiProvider.authenticateRequest(requestHeader) match {
+      case Authenticated(authedUser) => Right(authedUser)
+      case Invalid(message, throwable) => Left(unauthorised(message, throwable))
+      case NotAuthorised(message) => Left(forbidden(s"Principal not authorised: $message"))
+      case NotAuthenticated =>
+        providers.userProvider.authenticateRequest(requestHeader) match {
+          case NotAuthenticated => Left(unauthorised("Not authenticated"))
+          case Expired(principal) => Left(expired(principal))
+          case GracePeriod(authedUser) => Right(authedUser)
```

Comment:
Do we need `GracePeriod` in the provider API at all? From a quick glance at the code I think we could make it an internal concern to the Panda provider, but there might be something I missed